댓글 API 연동

WhiteHyun opened the pull request:
## 📌 PR 요약

🌱 작업한 내용
- API로 받아온 모델로 UI 연동

🌱 PR 포인트
- `UICollectionViewDiffableDataSource`를 사용하여 구현했습니다. RxDataSource와 유사하지만, Rx를 이용한 파싱 처리보다 간결하게 쓰이는 것 같았습니다(제 머릿속으로는 이게 더 깔끔했는데 아니라면 언제든 의견 주세요.. 제가 바보라서 못하는 걸 수도 있어요).
   - `DiffableDataSource`는 #17 때 설명드렸으므로 스킵하겠습니다. 자세한 설명을 더 듣고싶다면 언제든 슬랙이나 디코로 연락주세요.

## 📸 스크린샷
|스크린샷|
|:--:|
|<img src="https://user-images.githubusercontent.com/57972338/224964302-50abdec6-9cf1-4fda-9d30-9c5f4139c3a5.gif" width="40%"/>|
|View Hierarchy|
|<img src="https://user-images.githubusercontent.com/57972338/225183538-63a44276-666b-4d90-810b-face22905f03.png" width="40%"/>|

- 각 섹션마다 하나의 댓글을 가지도록 구성했으며, 댓글이 삭제되면 하나의 섹션을 삭제하는 식으로 구현할 예정입니다.
- 두 번째 섹션부터 댓글이 보이도록 설정한 이유는 게시글이 첫 번째 섹션 헤더로 고정되어있기 때문입니다. 만약 댓글이 존재하지 않는다면, 섹션 값은 `0`이기 때문에 게시물도 보이지 않게 됩니다. 따라서, 댓글이 없어도 게시글은 보일 수 있도록 상단에 섹션 하나를 고정해두었습니다.

## 📮 관련 이슈
- #211

WhiteHyun committed:
[ADD] FeedsContent conforms to `Hashable`

WhiteHyun committed:
[ADD] BoardDetailViewModel conforms to `BoardDetailDataStore`

WhiteHyun committed:
[ADD] CommentContent conforms to `Hashable`

WhiteHyun committed:
[CHORE] groupID's rawValue from CodingKeys

WhiteHyun committed:
[ADD] configure(with:) in BoardDetailCollectionViewCell

WhiteHyun committed:
[CHORE] collectionView header's height when section is not zero

WhiteHyun committed:
[CHORE] BoardDetailCollectionViewCell's configure(with:) methods

WhiteHyun committed:
[ADD] BoardDetailCollectionHeaderView's configure(with:)

WhiteHyun committed:
[FEAT] Feed detail navigation on model selection in collection view

BoardDetailViewController에게 View를 띄워주기 위해 미리 처리해야할 모델은
FeedsContent, [CommentContent] 타입의 두 가지 모델입니다.
ClipboardViewController에서 직접 API 요청을 한 뒤 받아온 값을 전달하면서
NavigationController에 Push하여 ClipboardVC에서는 미리 받아온 모델로
뷰를 구성할 수 있도록 처리했습니다.

WhiteHyun committed:
[DEL] UICollectionViewDataSource

DiffableDataSource로 이전하기 위한 전처리입니다.

WhiteHyun committed:
[ADD] Type alias for Diffable DataSource

WhiteHyun committed:
[FEAT] CollectionView Diffable Datasource Registration

WhiteHyun committed:
[FEAT] Snapshots for CollectionView initial state

WhiteHyun committed:
[CHORE] BoardDetailCollectionViewCell's dynamic height logic

viewModel이 갖고있는 comments에 접근하여 댓글의 정보를 갖고 있을 때
높이를 가지고 Size를 처리하도록 함

WhiteHyun committed:
[FIX] UI layout issue caused by unordered Set

WhiteHyun commented:
어우, 페이징 처리를 안했네요.. 다음 PR때 구현해서 올리겠습니다. 😅

WhiteHyun committed:
[FIX] Comment Cell's dynamic height

- 셀의 정보에 맞게 사이징 작업이 되어있지 않았습니다.
- 셀에 들어갈 모델에 따라 hidden처리 되는 UI도 있고, 텍스트도 다르기 때문에
  단순 댓글만이 아닌 모델 전체를 파라미터로 해야할 필요가 있었습니다.

soobin-k reviewed (approved):
복잡한 UI였는데,,너무 고생많으셨습니다 승현님!!
추후 내비바 back버튼도 텍스트 없애도록 수정하시면 좋을 것 같습니다!

WhiteHyun committed:
[REFACTOR] Fetching feed content and comment contents

피드에 진입 전에 API로 정보를 받아온 뒤 ViewController를 Push했지만,
API 처리 속도가 느린 경우 사용자 경험을 저해할 수 있습니다.
이에 이전 ViewController가 갖고있는 Feed content를 가지고, push된 ViewController가
직접 API를 요청하여 데이터를 처리하도록 구현했습니다.
